### PR TITLE
[SPARK-14029][SQL] Improve BooleanSimplification optimization by implementing `Not` canonicalization.

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

Currently, **BooleanSimplification** optimization can handle the following cases.
- a && (!a || b ) ==> a && b
- a && (b || !a ) ==> a && b

However, it can not handle the followings cases since those equations fail at the comparisons between their canonicalized forms.
- a < 1 && (!(a < 1) || b)     ==> (a < 1) && b
- a <= 1 && (!(a <= 1) || b) ==> (a <= 1) && b
- a > 1 && (!(a > 1) || b)     ==> (a > 1) && b
- a >= 1 && (!(a >= 1) || b) ==> (a >= 1) && b

This PR implements the above cases and also the followings, too.
- a < 1 && ((a >= 1) || b )   ==> (a < 1) && b
- a <= 1 && ((a > 1) || b )   ==> (a <= 1) && b
- a > 1 && ((a <= 1) || b)  ==> (a > 1) && b
- a >= 1 && ((a < 1) || b)  ==> (a >= 1) && b
## How was this patch tested?

Pass the Jenkins tests including new test cases in BooleanSimplicationSuite.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Hi, @rxin .
Could you review this PR when you have some time?

**Comment by rxin** [comment]
cc @marmbrus on whether we want this or not.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala [hunk 1]
```diff
@@ -71,6 +71,15 @@ object Canonicalize extends {
...
     case LessThanOrEqual(l, r) if l.hashCode() > r.hashCode() => GreaterThanOrEqual(r, l)
 
+    case Not(GreaterThan(l, r)) if l.hashCode() > r.hashCode() => GreaterThan(r, l)
+    case Not(GreaterThan(l, r)) => LessThanOrEqual(l, r)
+    case Not(LessThan(l, r)) if l.hashCode() > r.hashCode() => LessThan(r, l)
+    case Not(LessThan(l, r)) => GreaterThanOrEqual(l, r)
+    case Not(GreaterThanOrEqual(l, r)) if l.hashCode() > r.hashCode() => GreaterThanOrEqual(r, l)
+    case Not(GreaterThanOrEqual(l, r)) => LessThan(l, r)
+    case Not(LessThanOrEqual(l, r)) if l.hashCode() > r.hashCode() => LessThanOrEqual(r, l)
+    case Not(LessThanOrEqual(l, r)) => GreaterThan(l, r)
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala]: We should add some tests to `ExpressionSetSuite` for this.

Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala]: Thank you for review. I'll add more tests there.

**Comment by marmbrus** [comment]
Seems like a good idea to me

**Comment by dongjoon-hyun** [comment]
I added more testcases for `Not` canonicalization in ExpressionSetSuite.
Thank you, @rxin and @marmbrus .

**Comment by dongjoon-hyun** [comment]
Hi, @marmbrus . 
Could you review again for the newly added tests when you have a time? :)

**Comment by marmbrus** [comment]
LGTM, merging to master.

**Comment by dongjoon-hyun** [comment]
Oh, thank you so much for merging this PR, @marmbrus !

**Pull request closed without merging**